File: w5/emoticons.py
msg_str = input()
msg_set = set(msg_str)
emoticons = [":)", ":-)", ":-(", ";-)", "xD", "^_^", "-_-", "^o^", "^^;", "(..)"]
emo_set = set("".join(emoticons))
all_char_set = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ abcdefghijklmnopqrstuvwxyz!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"

def check_emoticons(msg):
    emoticons = [":)", ":-)", ":-(", ";-)", "xD", "^_^", "-_-", "^o^", "^^;", "(..)"]
    count = 0
    i = 0

    while i < len(msg):
        for emo in emoticons:
            emo_len = len(emo)
            if msg[i:i+emo_len] == emo:
                i += emo_len - 1
                break
        
        count += 1
        i += 1
    
    return count

max_count = 0
min_count = 101

for emo in emo_set:
    for char in msg_set:
        temp = msg_str.replace(char, emo)
        min_count = min(min_count, check_emoticons(temp))
        max_count = max(max_count, check_emoticons(temp))

print(min_count, max_count)

File: w5/emoticons.java
import java.util.*;

public class EmoticonChecker {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String msgStr = scanner.nextLine();
        Set<Character> msgSet = new HashSet<>();
        for (char c : msgStr.toCharArray()) {
            msgSet.add(c);
        }

        String[] emoticons = {":)", ":-)", ":-(", ";-)", "xD", "^_^", "-_-", "^o^", "^^;", "(..)"};
        Set<Character> emoSet = new HashSet<>();
        for (String emoticon : emoticons) {
            for (char c : emoticon.toCharArray()) {
                emoSet.add(c);
            }
        }

        int maxCount = 0;
        int minCount = 101;

        for (char emo : emoSet) {
            for (char charInMsg : msgSet) {
                String temp = msgStr.replace(charInMsg, emo);
                int count = checkEmoticons(temp, emoticons);
                minCount = Math.min(minCount, count);
                maxCount = Math.max(maxCount, count);
            }
        }

        System.out.println(minCount + " " + maxCount);
    }

    public static int checkEmoticons(String msg, String[] emoticons) {
        String temp = msg;
        for (String emoticon : emoticons) {
            temp = temp.replace(emoticon, "@"); // Replace emoticons with @
        }
        return temp.length();
    }
}
